[text structure] Find field and message structure endpoints (#105660)
elastic/elasticsearch: docs/reference/text-structure/apis/find-message-structure.asciidoc

[role="xpack"]
[[find-message-structure]]
= Find messages structure API

Finds the structure of a list of text messages.

[discrete]
[[find-message-structure-request]]
== {api-request-title}

`GET _text_structure/find_message_structure` +
`POST _text_structure/find_message_structure`

[discrete]
[[find-message-structure-prereqs]]
== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have `monitor_text_structure` or
`monitor` cluster privileges to use this API. See
<<security-privileges>>.

[discrete]
[[find-message-structure-desc]]
== {api-description-title}

This API provides a starting point for ingesting data into {es} in a format that

> is suitable for subsequent use with other {stack} functionality.

Reviewer: ```suggestion
is suitable for subsequent use with other {stack} functionality. Use this
API in preference to `findstructure` when your input text has already been
split up into separate messages by some other process.
```

include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-explain]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-format]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-grok-pattern]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-ecs-compatibility]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-quote]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-should-trim-fields]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timeout]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timestamp-field]
include::{es-repo-dir}/text-structure/apis/find-structure-shared.asciidoc[tag=param-timestamp-format]

[discrete]
[[find-message-structure-request-body]]
== {api-request-body-title}

`messages`::
(Required, array of strings)
The list of messages you want to analyze.

[discrete]
[[find-message-structure-examples]]
== {api-examples-title}

[discrete]
[[find-message-structure-example]]
=== Analyzing Elasticsearch log files

> Suppose you have a list of Elasticsearch logs messages.

Reviewer 2: ```suggestion
Suppose you have a list of {es} log messages.
```

You can send it to the `find_message_structure` endpoint as follows:

[source,console]
----
POST _text_structure/find_message_structure
{
  "messages": [
    "[2024-03-05T10:52:36,256][INFO ][o.a.l.u.VectorUtilPanamaProvider] [laptop] Java vector incubator API enabled; uses preferredBitSize=128",
    "[2024-03-05T10:52:41,038][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [repository-url]",
    "[2024-03-05T10:52:41,042][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [rest-root]",

>     "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]tive methods for [MacOS]",
>     "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]API enabled; uses preferredBitSize=128",

Reviewer: ```suggestion
    "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]",
    "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]",
```

          "timezone" : "{{ event.timezone }}",
          "formats" : [
            "ISO8601"
          ]
        }
      },
      {
        "remove" : {
          "field" : "timestamp"
        }
      }
    ]
  },
  "field_stats" : {
    "loglevel" : {
      "count" : 22,
      "cardinality" : 1,
      "top_hits" : [
        {
          "value" : "INFO",
          "count" : 22
        }
      ]
    },
    "message" : {
      "count" : 22,
      "cardinality" : 22,
      "top_hits" : [
        {
          "value" : "[2024-03-05T10:52:36,256][INFO ][o.a.l.u.VectorUtilPanamaProvider] [laptop] Java vector incubator API enabled; uses preferredBitSize=128",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,038][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [repository-url]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,042][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [rest-root]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [ingest-user-agent]",
          "count" : 1
        },
        {

>           "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]tive methods for [MacOS]",
>           "count" : 1
>         },
>         {
>           "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]API enabled; uses preferredBitSize=128",

Reviewer: ```suggestion
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-core]",
          "count" : 1
        },
        {
          "value" : "[2024-03-05T10:52:41,043][INFO ][o.e.p.PluginsService     ] [laptop] loaded module [x-pack-redact]",
```

          "count" : 1
        }
      ]
    },
    "timestamp" : {
      "count" : 22,
      "cardinality" : 14,
      "earliest" : "2024-03-05T10:52:36,256",
      "latest" : "2024-03-05T10:52:49,199",
      "top_hits" : [
        {
          "value" : "2024-03-05T10:52:41,044",
          "count" : 6
        },
        {
          "value" : "2024-03-05T10:52:41,043",
          "count" : 3
        },
        {
          "value" : "2024-03-05T10:52:41,059",
          "count" : 2
        },
        {
          "value" : "2024-03-05T10:52:36,256",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:41,038",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:41,042",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:43,291",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:46,098",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:47,227",
          "count" : 1
        },
        {
          "value" : "2024-03-05T10:52:47,259",
          "count" : 1
        }
      ]
    }
  }
}
----

> // TESTRESPONSE[s/"sample_start" : ".*",/"sample_start" : "$body.sample_start",/]
> // The substitution is because the text is pre-processed by the test harness,
> // so the fields may get reordered in the JSON the endpoint sees

Reviewer: ```suggestion
// TESTRESPONSE
```

In this case I don't see why the test harness would reorder the input. JSON arrays have a well-defined order.

For a detailed description of the response format, or for additional examples

> on ingesting delimited text (such as CSV) or newline-delimited JSON, see the

Reviewer 2: As above.
```suggestion
on ingesting delimited text (such as CSV) or newline-delimited JSON, refer to the
```
